### PR TITLE
Do not allow switching models if ECS is committing

### LINKED ISSUE #2971: Block switching models while deploying
When you create a new model in the deployment flow it's possible to then use the model switcher before the new model has been connected to, and the commit successfully completed.

This causes the commit to fail and the model to be in a weird state.

### PR DESCRIPTION
**Pull request by makyo**
Fixes #2971

### CODE REVIEW

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7958/

**Comment by makyo** [comment]
# QA

Add a bundle/lot of services. Click deploy and immediately try to switch models; you should get a notification.

**Comment by huwshimi** [comment]
@makyo Thanks for this!

Strangely, I clicked deploy and then immediately clicked "Start a new model", and I got the message to say "cannot switch models while deploying." but it did actually switch me to the new model!

File: jujugui/static/gui/src/app/utils/environment-change-set.js [hunk 1]
```diff
@@ -434,10 +436,20 @@ YUI.add('environment-change-set', function(Y) {
       this.currentCommit = this._buildHierarchy(true);
       this.currentIndex += 1;
       this.currentLevel = -1;
+      this.committing = true;
       this._commitNext(env, this.currentIndex - 1);
     },
 
     /**
+      Check to see if the current  commit is in progress.
```
Comment (frankban) [inline jujugui/static/gui/src/app/utils/environment-change-set.js]: Double space between "current" and "commit".

File: jujugui/static/gui/src/app/utils/environment-change-set.js [hunk 2]
```diff
@@ -374,6 +375,7 @@ YUI.add('environment-change-set', function(Y) {
           window.setTimeout(this._commitNext.bind(this, env, currentIndex), 0);
         } else {
           this.currentLevel = -1;
+          this.committing = false;
```
Comment (frankban) [inline jujugui/static/gui/src/app/utils/environment-change-set.js]: Nice, can we demonstrate that, sooner or later, in spite of errors, committing will be reset to false?
If so, could you please add a comment explaining why?

File: jujugui/static/gui/src/app/components/env-switcher/env-switcher.js [hunk 3]
```diff
@@ -116,7 +116,16 @@ class EnvSwitcher extends React.Component {
   handleModelClick(model) {
     const props = this.props;
     this.setState({showEnvList: false});
-    props.switchModel(model);
+    if (this.props.isCommitting()) {
+      const message = 'cannot switch models while deploying.';
```
Comment (frankban) [inline jujugui/static/gui/src/app/components/env-switcher/env-switcher.js]: s/cannot/Cannot/ ?

File: jujugui/static/gui/src/app/components/env-switcher/env-switcher.js [hunk 4]
```diff
@@ -116,7 +116,16 @@ class EnvSwitcher extends React.Component {
...
+      this.props.addNotification({
+        title: message,
+        message: message,
+        level: 'error'
+      });
```
Comment (frankban) [inline jujugui/static/gui/src/app/components/env-switcher/env-switcher.js]: Return here and avoid the else block? [shrug]

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7971/

**Comment by hatched** [comment]
Because `switchModel` here doesn't take a callback it could be expected to return a value indicating the success or failure of the call. Up until now we've simply been returning undefined when this failure case hits, but for now we want to indicate to the user why we aren't switching.

I propose that instead of passing the addNotification into here, which will create a notification separate from the users action, we return a boolean value from this function which the caller can use to update the user within it's own UI.

This would allow us to do things like 'cannot switch to already connected model' inline in the profile. Or a notification in the model switcher itself saying that it cannot switch because of x. I believe this also simplifies the codebase a bit because the notification support doesn't need to be included, or tested in `switchModel`

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7972/

File: jujugui/static/gui/src/app/utils/environment-change-set.js [hunk 5]
```diff
@@ -35,6 +35,7 @@ YUI.add('environment-change-set', function(Y) {
       this.changeSet = {};
       this.currentCommit = [];
       this.currentIndex = 0;
+      this.committing = false;
```
Comment (hatched) [inline jujugui/static/gui/src/app/utils/environment-change-set.js]: Can you add a test for this functionality plz

**Comment by hatched** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**